src/messenger.rs

```diff
@@ -922,6 +932,122 @@ mod tests {
+            let (mut tx_middle_read, mut tx_middle_write) = tokio::io::split(tx_middle);
+
+            let direction1 = async {
+                for i in 0.. {
+                    let mut buf = [0; 1];
+                    rx_middle_read.read_exact(&mut buf).await.unwrap();
+                    tx_middle_write.write_all(&buf).await.unwrap();
+
+                    if i == 3 {
+                        network_pause_captured.wait().await;
+                    }
+                }
+            };
+
+            let direction2 = async {
+                loop {
+                    let mut buf = [0; 1];
+                    tx_middle_read.read_exact(&mut buf).await.unwrap();
+                    rx_middle_write.write_all(&buf).await.unwrap();
+                }
+            };
+
+            tokio::select! {
+                _ = direction1 => {}
+                _ = direction2 => {}
+            }
+        });
+
+        // simulated broker, just reads messages and answers w/ "api versions" responses
+        let handle_broker = tokio::spawn(async move {
+            for correlation_id in 0.. {
+                rx_back.read_message(1_000).await.unwrap();
+
+                let mut msg = vec![];
+                ResponseHeader {
+                    correlation_id: Int32(correlation_id),
+                    tagged_fields: Default::default(), // NOT serialized for ApiVersion!
+                }
+                .write_versioned(&mut msg, ApiVersion(Int16(0)))
+                .unwrap();
+                let resp = ApiVersionsResponse {
+                    error_code: Some(ApiError::CorruptMessage),
+                    api_keys: vec![],
+                    throttle_time_ms: Some(Int32(1)),
+                    tagged_fields: Some(TaggedFields::default()),
+                };
+                resp.write_versioned(&mut msg, ApiVersionsRequest::API_VERSION_RANGE.min())
+                    .unwrap();
+                rx_back.write_message(&msg).await.unwrap();
+            }
+        });
+
+        messenger.set_version_ranges(HashMap::from([(
+            ApiKey::ApiVersions,
+            ApiVersionRange::new(ApiVersion(Int16(0)), ApiVersion(Int16(0))),
+        )]));
+
+        // send first message, this task will be canceled after 3 bytes got sent.
+        let task_to_cancel = (async {
+            messenger
+                .request(ApiVersionsRequest {
+                    client_software_name: CompactString(String::from("foo")),
+                    client_software_version: CompactString(String::from("bar")),
+                    tagged_fields: TaggedFields::default(),
+                })
+                .await
+                .unwrap();
+        })
+        .fuse();
+
+        {
+            // cancel when we exit this block
+            pin_mut!(task_to_cancel);
+
+            // write exactly 3 bytes via the client and then cancel the task.
+            futures::select_biased! {
```

Comment:
I think this is racy, as network_pause.wait() could be reached allowing the network to continue and the task to finish, before this task is scheduled and exits the select block. I'm not sure that select_biased will prevent this, as even if network_pause triggers the wakeup, task_to_cancel will still be polled first